### PR TITLE
Document `CALL { ... } IN TRANSACTIONS`

### PR DESCRIPTION
**Pull request by tobias-johansson**
I decided to put this as a subsection of the `CALL {} (subquery)` clause section, rather than as a top level clause section.

### CODE REVIEW

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc [hunk 1]
```diff
@@ -6,22 +6,31 @@
 This section describes how Cypher queries work with database transactions.
 --
 
-All Cypher statements are explicitly run within a transaction.
-For read-only queries, the transaction will always succeed.
-For updating queries it is possible that a failure can occur for some reason, for example if the query attempts to violate a constraint, in which case the entire transaction is rolled back, and no changes are made to the graph.
-Every statement is executed within the context of the transaction, and nothing will be persisted to disk until that transaction is successfully committed.
+All Cypher queries run within transactions. Modifications done by updating queries are held in memory by the transaction until it is committed, at which point the changes are persisted to disk and become visible to other transactions.
```
Comment (hedengran) [inline cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]: Missing newline after first sentence.

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc [hunk 2]
```diff
@@ -6,22 +6,31 @@
...
+
+Note that, as the transaction holds changes in memory, a query that makes a large number of updates will consequently use large amounts of memory.
+For memory configuration in Neo4j, see the <<operations-manual#memory-configuration, Neo4j Operations Manual -> Memory configuration>>.
+
+Transactions can be either _explicit_ or _implicit_.
+
+- _Explicit_ transactions,
+  * are opened by the user,
+  * can execute multiple Cypher queries in sequence,
+  * are committed, or rolled back, by the user
```
Comment (hedengran) [inline cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]: I think this might look better, but just my 2 cents. 

```suggestion
- _Explicit_ transactions:
  * are opened by the user.
  * can execute multiple Cypher queries in sequence.
  * are committed, or rolled back, by the user.
```

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc [hunk 3]
```diff
@@ -6,22 +6,31 @@
...
+
+- _Implicit_ transactions, sometimes called auto-commit transactions or auto transactions,
+  * are opened automatically,
+  * can execute a single Cypher query,
+  * are committed automatically when the query finishes successfully
```
Comment (hedengran) [inline cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]:
```suggestion
- _Implicit_ transactions, sometimes called auto-commit transactions or auto transactions:
  * are opened automatically.
  * can execute a single Cypher query.
  * are committed automatically when the query finishes successfully.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 4]
```diff
@@ -208,7 +212,91 @@ class CallSubqueryTest extends DocumentingTest {
...
+      p("""Subqueries can be made to execute in separate, inner transactions, producing intermediate commits.
+          #This can come in handy when doing large write operations, like batch updates or imports.
+          #To execute a subquery in separate transactions you add the modifier `IN TRANSACTIONS` after the subquery.""".stripMargin('#'))
+
+      implicit val csvFilesDir: File = createDir("csv-files")
+      val artistsCsvPath = new CsvFile("artists.csv").withContents(
+        Seq("1", "ABBA", "1992"),
+        Seq("2", "Roxette", "1986"),
+        Seq("3", "Europe", "1979"),
+        Seq("4", "The Cardigans", "1992")
```
Comment (hedengran) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]:
```suggestion
        Seq("4", "The Cardigans", "1992")
        Seq("5", "bob hund", "1991")
```

jk 😉

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 5]
```diff
@@ -208,7 +212,91 @@ class CallSubqueryTest extends DocumentingTest {
...
+        """LOAD CSV FROM '@csvFile' AS line
+          #CALL {
+          #  WITH line
+          #  CREATE (:Artist {name: line[1], year: toInteger(line[2])})
+          #} IN TRANSACTIONS OF 1 ROW""".stripMargin('#'),
+        assertions = ResultAssertions(r => r.isEmpty),
+        replacements = Seq(QueryTextReplacement("@csvFile", "file:///artists.csv", artistsCsvPath))
+      ) { resultTable() }
+      p("""As the size of the CSV file in this example is small, only a single separate transaction is
+          #started and committed.""".stripMargin('#'))
```
Comment (hedengran) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala:248]: Isn't four transactions started and committed? 😲

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 6]
```diff
@@ -208,7 +212,91 @@ class CallSubqueryTest extends DocumentingTest {
...
+            #} IN TRANSACTIONS OF 2 ROWS""".stripMargin('#'),
+          assertions = ResultAssertions(r => r.isEmpty),
+          replacements = Seq(QueryTextReplacement("@csvFile", "file:///artists.csv", artistsCsvPath))
+        ) { resultTable() }
+        p("""The query now starts and commits two separate transactions""")
+        p(""". The first two executions of the subquery (for the first two input rows from `LOAD CSV`) take place in the first transaction.
+            |. The first transaction is then committed before proceeding.
+            |. The next two executions of the subquery (for the next two input rows) take place in a second transaction.
+            |. The second transaction is committed.""".stripMargin)
+      }
```
Comment (hedengran) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]: Should we maybe include an example where `inputRows % batchSize != 0` just to clarify how we deplete the last batch?

**Review by hedengran — CHANGES_REQUESTED** [review]
Nice work! Just a few comments. Additionally:

* Should we include some example in the Cypher ref-card?
*  Might be a good time to add the deprecation of periodic commit to `Deprecations, additions and compatibility` as well.

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala (hunk 4 above)
Comment (tobias-johansson) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]: 🤣

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala (hunk 5 above)
Comment (tobias-johansson) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala:248]: Oops wasn't supposed to have a `OF 1 ROW` in the first example

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala (hunk 6 above)
Comment (tobias-johansson) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]: Yeah, that makes sense. Gives us a good reason to add Bob hund to the csv 👍

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 7]
```diff
@@ -224,7 +224,8 @@ class CallSubqueryTest extends DocumentingTest {
         Seq("1", "ABBA", "1992"),
         Seq("2", "Roxette", "1986"),
         Seq("3", "Europe", "1979"),
-        Seq("4", "The Cardigans", "1992")
+        Seq("4", "bob hund", "1991"),
```
Comment (hedengran) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala:227]: 😁

**Review by hedengran — APPROVED** [review]

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc (hunk 1 above)
Comment (tobias-johansson) [reply cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]: It will all be one contiguous paragraph anyway won't it?

Comment (hedengran) [reply cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]: True, but I think that for markup files it's good practice to have each sentence on a new line. It's easier to edit and `diffs` will be a lot easier to look at. This resource even quotes the writing guide for `Neo4j Documentation` (which I didn't know existed): https://asciidoctor.org/docs/asciidoc-recommended-practices/

File: cypher/cypher-docs/src/docs/dev/clause.asciidoc [hunk 8]
```diff
@@ -210,6 +210,9 @@ Duplicates are retained.
 m|<<query-call-subquery, CALL +++{ ... }+++>>
 |Evaluates a subquery, typically used for post-union processing or aggregations.
 
+m|<<subquery-call-in-transactions, CALL +++{ ... }+++ IN TRANSACTIONS>>
+|Evaluates a subquery in separate transactions. Typically used when modifying or importing large amounts of data .
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/clause.asciidoc]:
```suggestion
|Evaluates a subquery in separate transactions. 
Typically used when modifying or importing large amounts of data.
```

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc [hunk 9]
```diff
@@ -6,22 +6,28 @@
...
-All Cypher statements are explicitly run within a transaction.
-For read-only queries, the transaction will always succeed.
-For updating queries it is possible that a failure can occur for some reason, for example if the query attempts to violate a constraint, in which case the entire transaction is rolled back, and no changes are made to the graph.
-Every statement is executed within the context of the transaction, and nothing will be persisted to disk until that transaction is successfully committed.
+All Cypher queries run within transactions.
+Modifications done by updating queries are held in memory by the transaction until it is committed, at which point the changes are persisted to disk and become visible to other transactions.
+If an error occurs - either during query evaluation, such as division by zero, or during commit, such as constraint violations - the transaction is automatically rolled back, and no changes are persisted in the graph.
 
-In short, an updating query will always either fully succeed, or not succeed at all.
+In short, updating queries will always either fully succeed, or not succeed at all.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]:
```suggestion
In short, updating a query always either fully succeeds, or does not succeed at all.
```

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc [hunk 10]
```diff
@@ -6,22 +6,28 @@
...
 
-While it is not possible to run a Cypher query outside a transaction, it is possible to run multiple queries within a single transaction using the following sequence of operations:
+Note that, as the transaction holds changes in memory, a query that makes a large number of updates will consequently use large amounts of memory.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc]:
```suggestion
[NOTE]
====
A query that makes a large number of updates consequently uses large amounts of memory since the transaction holds changes in memory.
For memory configuration in Neo4j, see the <<operations-manual#memory-configuration, Neo4j Operations Manual -> Memory configuration>>.
====
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 11]
```diff
@@ -208,7 +212,95 @@ class CallSubqueryTest extends DocumentingTest {
...
+          """UNWIND [4, 2, 1, 0] AS i
+            #CALL {
+            #  WITH i
+            #  CREATE (:Example {num: 100/i})
+            #} IN TRANSACTIONS OF 2 ROWS
+            #RETURN i""".stripMargin('#'),
+          assertions = ErrorAssertions(t => t.getMessage should (include("/ by zero"))),
+          databaseStateBehavior = KeepState,
+        ) { errorOnlyResultTable() }
+        p("""When the failure occurred, the first transaction had already committed, so the database contains two example nodes""")
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]:
```suggestion
        p("""When the failure occurred, the first transaction had already been committed, so the database contains two example nodes""")
```

File: cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc [hunk 12]
```diff
@@ -6,22 +6,28 @@
...
+For memory configuration in Neo4j, see the <<operations-manual#memory-configuration, Neo4j Operations Manual -> Memory configuration>>.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/general/query-transactions.asciidoc:18]:
```suggestion

```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 13]
```diff
@@ -208,7 +212,95 @@ class CallSubqueryTest extends DocumentingTest {
...
+            |. The first transaction is then committed before proceeding.
+            |. The next two executions of the subquery (for the next two input rows) take place in a second transaction.
+            |. The second transaction is committed.
+            |. The last execution of the subquery (for the last input row) takes place in a third transaction.
+            |. The third transaction is committed.
+            |""".stripMargin)
+      }
+      section("Errors") {
+        p("""If an error occurs in `CALL {} IN TRANSACTIONS` the entire query fails,
+            |the current inner transaction, and the outer transaction are rolled back.""".stripMargin)
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]:
```suggestion
            |both the current inner transaction and the outer transaction are rolled back.""".stripMargin)
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 14]
```diff
@@ -208,7 +212,95 @@ class CallSubqueryTest extends DocumentingTest {
...
+        p(""". The first two executions of the subquery (for the first two input rows from `LOAD CSV`) take place in the first transaction.
+            |. The first transaction is then committed before proceeding.
+            |. The next two executions of the subquery (for the next two input rows) take place in a second transaction.
+            |. The second transaction is committed.
+            |. The last execution of the subquery (for the last input row) takes place in a third transaction.
+            |. The third transaction is committed.
+            |""".stripMargin)
+      }
+      section("Errors") {
+        p("""If an error occurs in `CALL {} IN TRANSACTIONS` the entire query fails,
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]:
```suggestion
        p("""If an error occurs in `CALL {} IN TRANSACTIONS` the entire query fails and
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala [hunk 15]
```diff
@@ -208,7 +212,95 @@ class CallSubqueryTest extends DocumentingTest {
...
+            |the current inner transaction, and the outer transaction are rolled back.""".stripMargin)
+        important(p("""On error, any previously committed inner transactions remain committed, and are not rolled back."""))
+        p("""In the following example, the last subquery execution, in the second inner transaction, fails
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/CallSubqueryTest.scala]:
```suggestion
        p("""In the following example, the last subquery execution in the second inner transaction fails
```

**Review by AlexicaWright — COMMENTED** [review]
Some editorial comments.

File: cypher/cypher-docs/src/docs/dev/clause.asciidoc [hunk 16]
```diff
@@ -211,7 +211,8 @@ m|<<query-call-subquery, CALL +++{ ... }+++>>
 |Evaluates a subquery, typically used for post-union processing or aggregations.
 
 m|<<subquery-call-in-transactions, CALL +++{ ... }+++ IN TRANSACTIONS>>
-|Evaluates a subquery in separate transactions. Typically used when modifying or importing large amounts of data .
+|Evaluates a subquery in separate transactions. 
+Typically used when modifying or importing large amounts of data.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/clause.asciidoc:215]: May need a pipe?

Comment (tobias-johansson) [reply cypher/cypher-docs/src/docs/dev/clause.asciidoc:215]: Looks ok in my intellij asciidoc preview

Comment (AlexicaWright) [reply cypher/cypher-docs/src/docs/dev/clause.asciidoc:215]: I'll trust that, so 👍 from me now!
